Use `GlobalRef` of `Core.CodeInfo` in `@generated`

devmotion opened the pull request:
This PR addresses https://github.com/JuliaLang/julia/pull/40778/files#r784416018 and replaces the reference to `Core.CodeInfo` in `@generated` with a `GlobalRef`.

The current implementation caused errors in Turing (https://github.com/TuringLang/Turing.jl/issues/1756) which collected core AD parts in  a `Core` submodule (not anymore, it's renamed to `Essential` now since it's not the first time this caused problems: https://github.com/simonster/Reexport.jl/pull/33#issuecomment-904564731).

Currently the PR is still missing some tests. It's a bit embarrassing but I did not manage to come up with a MWE that would be fixed by this PR. I tried things like
```julia
module M
module Core
@generated f(x) = :(2 * x)
end
end

M.Core.f(3)
```
but they all seemed to work fine. Maybe someone has a suggestion how I can reproduce the bug and ensure that `Core` in `@generated` refers to the submodule (or something else) without `CodeInfo`?

devmotion committed:
Use `GlobalRef` of `Core.CodeInfo` in `@generated`

simeonschaub commented on base/expr.jl:
```diff
@@ -582,7 +582,7 @@ macro generated(f)
                                    # https://github.com/JuliaLang/julia/issues/25678
                                    Expr(:block,
                                         :(local tmp = $body),
```
Looking at this again, this also shouldn't use `tmp` as a name without any escaping, since this example currently errors:
```julia
julia> @test @eval Module() begin
           @generated f(tmp) = tmp
           f(1) == Int
       end
Error During Test at REPL[4]:1
  Test threw exception
  Expression: #= REPL[4]:1 =# @eval Module() begin
        #= REPL[4]:2 =# @generated f(tmp) = begin
                    tmp
                end
        f(1) == Int
    end
  syntax: local variable name "tmp" conflicts with an argument
[...]
```
We should either escape only the relevant parts properly or use `gensym` here. (This change to using `GlobalRef` here still seems useful regardless.)

simeonschaub reviewed (commented):
Here's a test case that currently errors on nightly:
```julia
@test @eval Module() begin
    Core = 1
    @generated f() = 1
    f() == 1
end
```

devmotion commented:
:facepalm: I realized I couldn't come up with a MWE because I used Julia 1.7.1 instead of 1.6.5 or master.

devmotion committed:
Fix clash with argument named `tmp` and add tests

Co-authored-by: Simeon Schaub <schaub@mit.edu>

devmotion commented:
I added a fix for arguments of name `tmp` as well and added both tests (added you as a commit co-author @simeonschaub). I used `gensym` since it seemed easier.

simeonschaub commented on base/expr.jl:
```diff
@@ -574,15 +574,16 @@ macro generated(f)
     if isa(f, Expr) && (f.head === :function || is_short_function_def(f))
         body = f.args[2]
         lno = body.args[1]
+        tmp = gensym("tmp")
```
Just a little shorter:
```suggestion
        @gensym tmp
```

devmotion replied:
Actually I used this first but got a compilation error. Maybe I just have to run a `make clean` though, I'll try again.

simeonschaub replied:
A wait, this might just be an issue with `@gensym` not being defined yet during bootstrap. In that case just leave it like this. It doesn't save much anyways.

devmotion replied:
Compilation still fails with the following error message:
```shell
error during bootstrap:
LoadError(at "compiler/compiler.jl" line 3: LoadError(at "expr.jl" line 544: LoadError(at "expr.jl" line 577: UndefVarError(var=:error))))
ijl_undefined_var_error at /home/david/sources/julia/src/rtutils.c:132
ijl_get_binding_or_error at /home/david/sources/julia/src/module.c:387
unknown function (ip: 0x7f426f913ed5)
unknown function (ip: 0x7f426f913e33)
unknown function (ip: 0x7f426f913ac4)
unknown function (ip: 0x7f426f9139e4)
unknown function (ip: 0x7f426f913237)
jl_invoke_julia_macro at /home/david/sources/julia/src/ast.c:1018
jl_expand_macros at /home/david/sources/julia/src/ast.c:1074
jl_expand_macros at /home/david/sources/julia/src/ast.c:1117
jl_expand_macros at /home/david/sources/julia/src/ast.c:1117
jl_expand_macros at /home/david/sources/julia/src/ast.c:1117
jl_expand_macros at /home/david/sources/julia/src/ast.c:1117
jl_expand_macros at /home/david/sources/julia/src/ast.c:1084
ijl_expand_with_loc_warn at /home/david/sources/julia/src/ast.c:1180
jl_parse_eval_all at /home/david/sources/julia/src/toplevel.c:1019
ijl_load_ at /home/david/sources/julia/src/toplevel.c:1069
unknown function (ip: 0x7f426f91083d)
unknown function (ip: 0x7f426f910784)
jl_apply at /home/david/sources/julia/src/julia.h:1777 [inlined]
do_call at /home/david/sources/julia/src/interpreter.c:126
eval_value at /home/david/sources/julia/src/interpreter.c:215
eval_stmt_value at /home/david/sources/julia/src/interpreter.c:166 [inlined]
eval_body at /home/david/sources/julia/src/interpreter.c:612
jl_interpret_toplevel_thunk at /home/david/sources/julia/src/interpreter.c:744
top-level scope at compiler/compiler.jl:46
jl_toplevel_eval_flex at /home/david/sources/julia/src/toplevel.c:888
jl_eval_module_expr at /home/david/sources/julia/src/toplevel.c:195 [inlined]
jl_toplevel_eval_flex at /home/david/sources/julia/src/toplevel.c:675
ijl_toplevel_eval_in at /home/david/sources/julia/src/toplevel.c:947
unknown function (ip: 0x7f426f90fc0d)
jl_apply at /home/david/sources/julia/src/julia.h:1777 [inlined]
do_call at /home/david/sources/julia/src/interpreter.c:126
eval_value at /home/david/sources/julia/src/interpreter.c:215
eval_stmt_value at /home/david/sources/julia/src/interpreter.c:166 [inlined]
eval_body at /home/david/sources/julia/src/interpreter.c:612
jl_interpret_toplevel_thunk at /home/david/sources/julia/src/interpreter.c:744
top-level scope at compiler/compiler.jl:3
jl_toplevel_eval_flex at /home/david/sources/julia/src/toplevel.c:888
jl_parse_eval_all at /home/david/sources/julia/src/toplevel.c:1022
ijl_load_ at /home/david/sources/julia/src/toplevel.c:1069
ijl_load at /home/david/sources/julia/src/toplevel.c:1082
exec_program at /home/david/sources/julia/src/jlapi.c:521
true_main at /home/david/sources/julia/src/jlapi.c:574
jl_repl_entrypoint at /home/david/sources/julia/src/jlapi.c:706
main at /home/david/sources/julia/cli/loader_exe.c:59
__libc_start_main at /usr/lib/libc.so.6 (unknown line)
_start at /home/david/sources/julia/usr/bin/julia (unknown line)

make[1]: *** [sysimage.mk:61: /home/david/sources/julia/usr/lib/julia/corecompiler.ji] Error 1
make: *** [Makefile:82: julia-sysimg-ji] Error 2
```
Any idea if and how this can be fixed?

devmotion replied:
Ah just saw your message now :slightly_smiling_face: OK, I'll leave it.

simeonschaub replied:
Yeah, this is an issue with bootstrapping. We define `Core.Compiler` and `Base` bit by bit, so if code gets executed earlier in the process, you don't have access to all higher-level functions and macros yet.

simeonschaub reviewed (commented):
Thanks! That's a great first contribution!

simeonschaub reviewed (approved):
Ah, I actually meant to approve this...
